File: apps/server-admin/src/lib/system-notifications/providers/dingtalk.ts
import { createHmac } from "node:crypto";
import type {
  NotificationDispatchContext,
  NotificationMessage,
  NotificationProvider,
  NotificationProviderDefinition,
  NotificationSchemaField,
  NotificationSendResult,
} from "../types";
import {
  splitCommaSeparatedValues,
  toPlainRecord,
  toTrimmedString,
  truncateText,
} from "./shared";
import { tDefault } from "../../i18n";

const dingtalkT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) =>
  tDefault(`server.notifications.providers.catalog.dingtalk.${key}`, params);

const DINGTALK_CONNECTION_SCHEMA: NotificationSchemaField[] = [
  {
    key: "webhook_url",
    label: "Webhook URL",
    description: dingtalkT("fields.webhook_url.description"),
    placeholder: "https://oapi.dingtalk.com/robot/send?access_token=xxxxxx",
    type: "string",
    required: true,
    sensitive: true,
  },
  {
    key: "secret",
    label: dingtalkT("fields.secret.label"),
    description: dingtalkT("fields.secret.description"),
    placeholder: "SECxxxxxxxx",
    type: "string",
    sensitive: true,
  },
  {
    key: "keyword_prefix",
    label: dingtalkT("fields.keyword_prefix.label"),
    description: dingtalkT("fields.keyword_prefix.description"),
    placeholder: dingtalkT("fields.keyword_prefix.placeholder"),
    type: "string",
  },
  {
    key: "timeout_seconds",
    label: dingtalkT("fields.timeout_seconds.label"),
    type: "number",
    required: true,
    default_value: 5,
    min: 1,
    max: 30,
  },
];

const DINGTALK_TARGET_SCHEMA: NotificationSchemaField[] = [
  {
    key: "at_mobiles",
    label: dingtalkT("fields.at_mobiles.label"),
    description: dingtalkT("fields.at_mobiles.description"),
    placeholder: "13800001111,13900002222",
    type: "string",
  },
  {
    key: "at_user_ids",
    label: dingtalkT("fields.at_user_ids.label"),
    description: dingtalkT("fields.at_user_ids.description"),
    placeholder: "manager7675,user123",
    type: "string",
  },
  {
    key: "is_at_all",
    label: dingtalkT("fields.is_at_all.label"),
    description: dingtalkT("fields.is_at_all.description"),
    type: "boolean",
    default_value: false,
  },
];

export const dingtalkProviderDefinition: NotificationProviderDefinition = {
  type: "dingtalk",
  label: dingtalkT("label"),
  description: dingtalkT("description"),
  connection_schema: DINGTALK_CONNECTION_SCHEMA,
  target_schema: DINGTALK_TARGET_SCHEMA,
  sensitive_fields: ["webhook_url", "secret"],
  capabilities: {
    supports_text: true,
    supports_markdown: true,
    supports_rich_blocks: false,
    supports_actions: true,
    supports_mentions: true,
    supports_attachments: false,
    supports_provider_dedupe_key: false,
    max_body_length: null,
  },
};

const applyKeywordPrefix = (value: string, keyword: string) => {
  const trimmedKeyword = keyword.trim();
  const trimmedValue = value.trim();
  if (!trimmedKeyword) return trimmedValue;
  if (trimmedValue.includes(trimmedKeyword)) return trimmedValue;
  return trimmedValue
    ? `【${trimmedKeyword}】 ${trimmedValue}`
    : trimmedKeyword;
};

const buildQueryUrl = (url: string, params: Record<string, string>) => {
  try {
    const parsed = new URL(url);
    for (const [key, value] of Object.entries(params)) {
      parsed.searchParams.set(key, value);
    }
    return parsed.toString();
  } catch {
    const query = Object.entries(params)
      .map(
        ([key, value]) =>
          `${encodeURIComponent(key)}=${encodeURIComponent(value)}`,
      )
      .join("&");
    return `${url}${url.includes("?") ? "&" : "?"}${query}`;
  }
};

const redactDingTalkWebhookUrl = (value: string) => {
  try {
    const url = new URL(value);
    if (url.searchParams.has("access_token")) {
      url.searchParams.set("access_token", "<redacted>");
    }
    if (url.searchParams.has("sign")) {
      url.searchParams.set("sign", "<redacted>");
    }
    return url.toString();
  } catch {
    return value
      .replace(/access_token=[^&]+/gi, "access_token=<redacted>")
      .replace(/sign=[^&]+/gi, "sign=<redacted>");
  }
};

const buildDingTalkMentionText = (
  atMobiles: string[],
  atUserIds: string[],
  isAtAll: boolean,
) => {
  const tokens = [
    ...atMobiles.map((mobile) => `@${mobile}`),
    ...atUserIds.map((userId) => `@${userId}`),
  ];

  if (isAtAll) {
    tokens.unshift(dingtalkT("mentionAll"));
  }

  return tokens.join(" ").trim();
};

const buildDingTalkMarkdownText = (
  message: NotificationMessage,
  mentionText: string,
) => {
  const sections: string[] = [];
  const summary = toTrimmedString(message.summary);
  const bodySource =
    toTrimmedString(message.body_markdown) ||
    toTrimmedString(message.body_text);

  if (summary) {
    sections.push(summary);
  }
  if (bodySource) {
    sections.push(
      bodySource
        .split("\n")
        .map((line) => line.trim())
        .filter(Boolean)
        .join("\n"),
    );
  }
  if (message.facts.length > 0) {
    sections.push(
      message.facts
        .map((fact) => `- **${fact.label}**：${fact.value}`)
        .join("\n"),
    );
  }
  if (message.actions.length > 0) {
    sections.push(
      message.actions
        .filter(
          (action) =>
            toTrimmedString(action.label) && toTrimmedString(action.url),
        )
        .map((action) => `- [${action.label.trim()}](${action.url.trim()})`)
        .join("\n"),
    );
  }
  if (mentionText) {
    sections.push(mentionText);
  }

  return sections.filter(Boolean).join("\n\n");
};

export const sendDingTalkMessage = async (args: {
  provider: NotificationProvider;
  message: NotificationMessage;
  context?: Partial<NotificationDispatchContext>;
  timeoutSeconds: number;
}): Promise<NotificationSendResult> => {
  const webhookUrl = toTrimmedString(
    args.provider.connection_config.webhook_url,
  );
  if (!webhookUrl) {
    return {
      success: false,
      retryable: false,
      reason: dingtalkT("errors.missingWebhookUrl"),
    };
  }

  const secret = toTrimmedString(args.provider.connection_config.secret);
  const keywordPrefix = toTrimmedString(
    args.provider.connection_config.keyword_prefix,
  );
  const targetConfig = toPlainRecord(args.context?.target?.target_config);
  const atMobiles = splitCommaSeparatedValues(targetConfig.at_mobiles);
  const atUserIds = splitCommaSeparatedValues(targetConfig.at_user_ids);
  const isAtAll = Boolean(targetConfig.is_at_all);
  const mentionText = buildDingTalkMentionText(atMobiles, atUserIds, isAtAll);
  const title = applyKeywordPrefix(
    toTrimmedString(args.message.title || dingtalkT("message.fallbackTitle")),
    keywordPrefix,
  );
  const markdownText =
    buildDingTalkMarkdownText(args.message, mentionText) ||
    toTrimmedString(args.message.summary) ||
    title;

  const timestamp = secret ? String(Date.now()) : "";
  const sign = secret
    ? createHmac("sha256", secret)
        .update(`${timestamp}\n${secret}`, "utf8")
        .digest("base64")
    : "";
  const requestUrl =
    secret && timestamp && sign
      ? buildQueryUrl(webhookUrl, { timestamp, sign })
      : webhookUrl;
  const requestBody = {
    msgtype: "markdown",
    markdown: {
      title,
      text: markdownText,
    },
    ...(atMobiles.length > 0 || atUserIds.length > 0 || isAtAll
      ? {
          at: {
            ...(atMobiles.length > 0 ? { atMobiles } : {}),
            ...(atUserIds.length > 0 ? { atUserIds } : {}),
            isAtAll,
          },
        }
      : {}),
  };

  const controller = new AbortController();
  const timeout = setTimeout(
    () => controller.abort(),
    Math.max(1, args.timeoutSeconds) * 1000,
  );

  try {
    const response = await fetch(requestUrl, {
      method: "POST",
      headers: {
        "content-type": "application/json; charset=utf-8",
      },
      body: JSON.stringify(requestBody),
      signal: controller.signal,
    });
    const responseText = await response.text().catch(() => "");
    let parsedResponse: {
      errcode?: number;
      errmsg?: string;
    } | null = null;
    try {
      parsedResponse = responseText ? JSON.parse(responseText) : null;
    } catch {
      parsedResponse = null;
    }

    const apiSucceeded = response.ok && (parsedResponse?.errcode ?? 0) === 0;

    return {
      success: apiSucceeded,
      retryable:
        !apiSucceeded && (response.status >= 500 || response.status === 429),
      reason: apiSucceeded
        ? undefined
        : parsedResponse?.errmsg ||
          dingtalkT("errors.requestReturned", { status: response.status }),
      request_summary: {
        method: "POST",
        url: redactDingTalkWebhookUrl(requestUrl),
        msgtype: requestBody.msgtype,
        signed: Boolean(secret),
        mentioned_mobile_count: atMobiles.length,
        mentioned_user_count: atUserIds.length,
        is_at_all: isAtAll,
        title_preview: truncateText(title, 120),
      },
      response_summary: {
        status: response.status,
        ok: response.ok,
        errcode: parsedResponse?.errcode,
        errmsg: parsedResponse?.errmsg,
        body_preview: truncateText(responseText),
      },
    };
  } catch (error) {
    return {
      success: false,
      retryable: true,
      reason:
        error instanceof Error
          ? error.message
          : dingtalkT("errors.requestFailed"),
      request_summary: {
        method: "POST",
        url: redactDingTalkWebhookUrl(requestUrl),
        msgtype: requestBody.msgtype,
        signed: Boolean(secret),
      },
    };
  } finally {
    clearTimeout(timeout);
  }
};

File: apps/server-admin-rs/src/notifications/routes/providers/dingtalk.rs
use super::*;

pub(in crate::notifications::routes) fn dingtalk_definition() -> ProviderDefinition {
    webhook_like_definition(
        "dingtalk",
        "DingTalk",
        "Send notifications through DingTalk robot webhook.",
        &["webhook_url", "secret"],
        vec![
            string_schema("at_mobiles", "At mobiles", false, false, None)
                .placeholder("13800001111,13900002222"),
            string_schema("at_user_ids", "At user IDs", false, false, None)
                .placeholder("manager7675,user123"),
            bool_schema("is_at_all", "At all", false, Some(false)),
        ],
    )
}

pub(in crate::notifications::routes) async fn send_dingtalk(
    state: &AppState,
    provider: &Value,
    target: &Value,
    message: &Value,
    timeout_seconds: i64,
) -> ProviderTestResult {
    let config = provider_config(provider);
    let target_config = target_config(target);
    let webhook_url = config_text(&config, "webhook_url");
    if webhook_url.is_empty() {
        return missing_config_result(&notification_provider_error_default(
            "dingtalk",
            "missingWebhookUrl",
            &[],
        ));
    }
    let secret = config_text(&config, "secret");
    let keyword_prefix = config_text(&config, "keyword_prefix");
    let at_mobiles = split_values(target_config.get("at_mobiles"));
    let at_user_ids = split_values(target_config.get("at_user_ids"));
    let is_at_all = target_config
        .get("is_at_all")
        .map(value_to_bool)
        .unwrap_or(false);
    let mention_text = build_dingtalk_mention_text(&at_mobiles, &at_user_ids, is_at_all);
    let title = apply_keyword_prefix(&message_title(message), &keyword_prefix);
    let markdown_text = non_empty_or(
        build_markdown_body(message, &mention_text),
        message_summary(message),
        &title,
    );
    let request_url = if secret.is_empty() {
        webhook_url.clone()
    } else {
        let timestamp = time_utils::now_ms().to_string();
        let sign = hmac_sha256_base64(
            secret.as_bytes(),
            format!("{timestamp}\n{secret}").as_bytes(),
        );
        append_query_params(&webhook_url, &[("timestamp", timestamp), ("sign", sign)])
    };
    let body = json!({
        "msgtype": "markdown",
        "markdown": { "title": title, "text": markdown_text },
        "at": {
            "atMobiles": at_mobiles,
            "atUserIds": at_user_ids,
            "isAtAll": is_at_all
        }
    });
    let request_summary = json!({
        "method": "POST",
        "url": redact_query_value(&redact_query_value(&request_url, "access_token"), "sign"),
        "msgtype": "markdown",
        "signed": !secret.is_empty(),
        "mentioned_mobile_count": body.pointer("/at/atMobiles").and_then(Value::as_array).map(Vec::len).unwrap_or(0),
        "mentioned_user_count": body.pointer("/at/atUserIds").and_then(Value::as_array).map(Vec::len).unwrap_or(0),
        "is_at_all": is_at_all,
        "title_preview": truncate_text(body.pointer("/markdown/title").and_then(Value::as_str).unwrap_or(""), 120)
    });
    let (status, ok, text, parsed) = post_json(state, &request_url, &body, timeout_seconds).await;
    provider_result_from_api(
        "DingTalk",
        request_summary,
        status,
        ok,
        text,
        parsed,
        |value| json_i64(value, "errcode").unwrap_or(0) == 0,
        |value| json_text(value, "errmsg"),
    )
}
